Exception thrown in record compact constructor causes test failure

**Describe the bug**
Exceptions thrown in the compact constructor of a record causes EqualsVerifier to fail the test.

**To Reproduce**
```
public record Foo(String data) {
    public Foo {
        if (data.length() <= 3) {
            throw new IllegalStateException(data);
        }
    }
}
```

In my test: `EqualsVerifier.forClass(Foo.class).suppress(Warning.NULL_FIELDS).verify();`

The compact constructor is the result of converting a regular Java class:

```
public Foo(String data) {
    if (data.length() <= 3) {
        throw new IllegalStateException(data);
    }
    this.data = data;
}
```

(private final field, getter, generated equals/hashCode)

**Error message**
```
EqualsVerifier found a problem in class somewhere.Foo.
-> Record: failed to invoke constructor.

For more information, go to: https://www.jqno.nl/equalsverifier/errormessages
java.lang.AssertionError: EqualsVerifier found a problem in class somewhere.Foo.
-> Record: failed to invoke constructor.

For more information, go to: https://www.jqno.nl/equalsverifier/errormessages
	at nl.jqno.equalsverifier.api.SingleTypeEqualsVerifierApi.verify(SingleTypeEqualsVerifierApi.java:316)
	at somewhere.FooTest.name(FooTest.java:10)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:108)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:96)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at jdk.proxy2/jdk.proxy2.$Proxy5.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: nl.jqno.equalsverifier.internal.exceptions.ReflectionException: Record: failed to invoke constructor.
	at app//nl.jqno.equalsverifier.internal.util.Rethrow.rethrow(Rethrow.java:34)
	at app//nl.jqno.equalsverifier.internal.reflection.RecordObjectAccessor.callRecordConstructor(RecordObjectAccessor.java:140)
	at app//nl.jqno.equalsverifier.internal.reflection.RecordObjectAccessor.makeAccessor(RecordObjectAccessor.java:119)
	at app//nl.jqno.equalsverifier.internal.reflection.RecordObjectAccessor.scramble(RecordObjectAccessor.java:61)
	at app//nl.jqno.equalsverifier.internal.reflection.ClassAccessor.getRedAccessor(ClassAccessor.java:179)
	at app//nl.jqno.equalsverifier.internal.reflection.ClassAccessor.getRedObject(ClassAccessor.java:168)
	at app//nl.jqno.equalsverifier.internal.util.Configuration.ensureUnequalExamples(Configuration.java:199)
	at app//nl.jqno.equalsverifier.internal.util.Configuration.build(Configuration.java:99)
	at app//nl.jqno.equalsverifier.api.SingleTypeEqualsVerifierApi.buildConfig(SingleTypeEqualsVerifierApi.java:381)
	at app//nl.jqno.equalsverifier.api.SingleTypeEqualsVerifierApi.performVerification(SingleTypeEqualsVerifierApi.java:367)
	at app//nl.jqno.equalsverifier.api.SingleTypeEqualsVerifierApi.verify(SingleTypeEqualsVerifierApi.java:312)
	... 84 more
Caused by: java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:480)
	at nl.jqno.equalsverifier.internal.reflection.RecordObjectAccessor.lambda$callRecordConstructor$4(RecordObjectAccessor.java:141)
	at nl.jqno.equalsverifier.internal.util.Rethrow.rethrow(Rethrow.java:30)
	... 94 more
Caused by: java.lang.IllegalStateException: one
	at somewhere.Foo.<init>(Foo.java:6)
	... 101 more
```

**Expected behavior**
I expect EqualsVerifier to pass the test. In general, I expect the same behaviour for classes and equivalent records.

**Version**
3.9.1

**Additional context**
Java 17

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -13,6 +13,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 <a name="3.x"/>
 
 ## [Unreleased]
+### Fixed
+- Support records that don't allow 0 values in their fields, with Warning.ZERO_FIELDS. ([Issue 613](https://github.com/jqno/equalsverifier/issues/613))
 
 ## [3.9.1] - 2022-03-19
 ### Fixed

--- a/docs/_errormessages/record-failed-to-invoke-constructor.md
+++ b/docs/_errormessages/record-failed-to-invoke-constructor.md
@@ -1,0 +1,32 @@
+---
+title: Record: failed to invoke constructor
+---
+    Record: failed to invoke constructor
+    If the record does not accept 0 as a value for its fields, consider providing valid prefab values for those fields and suppressing Warning.ZERO_FIELDS.
+
+Sometimes, a constructor needs to validate its input, throwing an exception on certain values. This can lead to problems when the value `0` is not allowed, and the constructor is part of a record instead of a regular class, like this:
+
+{% highlight java %}
+public record Foo(int i) {
+    public Foo {
+        if (i < 42) {
+            throw new IllegalArgumentException();
+        }
+    }
+}
+{% endhighlight %}
+
+One of the things EqualsVerifier does, is to run checks with fields set to their default values. For reference types, the default value is `null`, which has its own [chapter in the manual](/equalsverifier/manual/null). For primitive types, the default value is `0` (or `0.0`, `\u0000`, `false`). In regular classes, EqualsVerifier bypasses the constructor, so the exception can never be thrown. However, reflection support is much more limited for records, and their constructors cannot be bypassed.
+
+Therefore, we need to signal EqualsVerifier to skip the checks with default values, by suppressing `Warning.ZERO_FIELDS`.
+
+Another thing EqualsVerifier does, is to run checks with fields set to certain prefab values. For integral types, these values are `1` and `2`. If these values are not allowed by the record's constructor either, we need to provide new prefab values as well.
+
+In these cases, the call to EqualsVerifier will look like this:
+
+{% highlight java %}
+EqualsVerifier.forClass(Foo.class)
+        .suppress(Warning.ZERO_FIELDS)
+        .withPrefabValues(int.class, 42, 1337)
+        .verify();
+{% endhighlight %}

--- a/equalsverifier-16/src/test/java/nl/jqno/equalsverifier/integration/extended_contract/RecordsTest.java
+++ b/equalsverifier-16/src/test/java/nl/jqno/equalsverifier/integration/extended_contract/RecordsTest.java
@@ -51,6 +51,7 @@ public class RecordsTest {
         EqualsVerifier
             .forClass(ValueCheckingRecord.class)
             .withPrefabValues(int.class, 10, 11)
+            .suppress(Warning.ZERO_FIELDS)
             .verify();
     }
 

--- a/equalsverifier-16/src/test/java/nl/jqno/equalsverifier/integration/extended_contract/RecordsTest.java
+++ b/equalsverifier-16/src/test/java/nl/jqno/equalsverifier/integration/extended_contract/RecordsTest.java
@@ -35,6 +35,19 @@ public class RecordsTest {
     }
 
     @Test
+    public void fail_whenConstructorChecksValue() {
+        EqualsVerifier.forClass(ValueCheckingRecord.class).verify();
+    }
+
+    @Test
+    public void succeed_whenConstructorChecksValue_givenPrefabValues() {
+        EqualsVerifier
+            .forClass(ValueCheckingRecord.class)
+            .withPrefabValues(int.class, 10, 11)
+            .verify();
+    }
+
+    @Test
     public void fail_whenRecordInvariantIsViolated_givenIntFieldIsModifiedInConstructor() {
         ExpectedException
             .when(() -> EqualsVerifier.forClass(BrokenInvariantIntFieldRecord.class).verify())
@@ -123,6 +136,14 @@ public class RecordsTest {
     record NullCheckingRecord(String s) {
         public NullCheckingRecord {
             Objects.requireNonNull(s);
+        }
+    }
+
+    record ValueCheckingRecord(int i) {
+        public ValueCheckingRecord {
+            if (i <= 3) {
+                throw new IllegalStateException("Error: expected " + i + " > 3");
+            }
         }
     }
 

--- a/equalsverifier-16/src/test/java/nl/jqno/equalsverifier/integration/extended_contract/RecordsTest.java
+++ b/equalsverifier-16/src/test/java/nl/jqno/equalsverifier/integration/extended_contract/RecordsTest.java
@@ -36,7 +36,14 @@ public class RecordsTest {
 
     @Test
     public void fail_whenConstructorChecksValue() {
-        EqualsVerifier.forClass(ValueCheckingRecord.class).verify();
+        ExpectedException
+            .when(() -> EqualsVerifier.forClass(ValueCheckingRecord.class).verify())
+            .assertFailure()
+            .assertMessageContains(
+                "Record:",
+                "failed to invoke constructor",
+                "Warning.ZERO_FIELDS"
+            );
     }
 
     @Test

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/Warning.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/Warning.java
@@ -181,5 +181,15 @@ public enum Warning {
      * <p>{@code EqualsVerifier} checks for this by default but it can be disabled by suppressing
      * this warning.
      */
-    BIGDECIMAL_EQUALITY
+    BIGDECIMAL_EQUALITY,
+
+    /**
+     * Disables checks with fields with value 0 within {@code equals}, {@code hashCode} and
+     * {@code toString} methods.
+     *
+     * <p>Sometimes the constructor of a class makes sure no field can be 0. If this is the case,
+     * and if the fields cannot be made 0 later in the lifecycle of the class by setters or other
+     * methods, suppress this warning to disable the checks with fields that have value 0.
+     */
+    ZERO_FIELDS
 }

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/FieldInspector.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/FieldInspector.java
@@ -31,6 +31,8 @@ public class FieldInspector<T> {
     }
 
     public void checkWithNull(
+        boolean isNullWarningSuppressed,
+        boolean isZeroWarningSuppressed,
         Set<String> nonnullFields,
         AnnotationCache annotationCache,
         FieldCheck<T> check
@@ -38,13 +40,15 @@ public class FieldInspector<T> {
         for (Field field : FieldIterable.of(classAccessor.getType())) {
             ObjectAccessor<T> reference = classAccessor.getDefaultValuesAccessor(
                 typeTag,
-                false,
+                isNullWarningSuppressed,
+                isZeroWarningSuppressed,
                 nonnullFields,
                 annotationCache
             );
             ObjectAccessor<T> changed = classAccessor.getDefaultValuesAccessor(
                 typeTag,
-                false,
+                isNullWarningSuppressed,
+                isZeroWarningSuppressed,
                 nonnullFields,
                 annotationCache
             );

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/FieldsChecker.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/FieldsChecker.java
@@ -76,13 +76,13 @@ public class FieldsChecker<T> implements Checker {
         inspector.check(symmetryFieldCheck);
         inspector.check(transitivityFieldCheck);
 
-        if (!config.getWarningsToSuppress().contains(Warning.NULL_FIELDS)) {
-            inspector.checkWithNull(
-                config.getNonnullFields(),
-                config.getAnnotationCache(),
-                skippingSignificantFieldCheck
-            );
-        }
+        inspector.checkWithNull(
+            config.getWarningsToSuppress().contains(Warning.NULL_FIELDS),
+            config.getWarningsToSuppress().contains(Warning.ZERO_FIELDS),
+            config.getNonnullFields(),
+            config.getAnnotationCache(),
+            skippingSignificantFieldCheck
+        );
 
         if (!config.getWarningsToSuppress().contains(Warning.BIGDECIMAL_EQUALITY)) {
             inspector.check(bigDecimalFieldCheck);

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/HierarchyChecker.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/HierarchyChecker.java
@@ -87,6 +87,7 @@ public class HierarchyChecker<T> implements Checker {
                 classAccessor.getDefaultValuesAccessor(
                     typeTag,
                     config.getWarningsToSuppress().contains(Warning.NULL_FIELDS),
+                    config.getWarningsToSuppress().contains(Warning.ZERO_FIELDS),
                     config.getNonnullFields(),
                     config.getAnnotationCache()
                 )

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/RecordChecker.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/RecordChecker.java
@@ -36,6 +36,7 @@ public class RecordChecker<T> implements Checker {
             accessor.getDefaultValuesAccessor(
                 config.getTypeTag(),
                 config.getWarningsToSuppress().contains(Warning.NULL_FIELDS),
+                config.getWarningsToSuppress().contains(Warning.ZERO_FIELDS),
                 config.getNonnullFields(),
                 config.getAnnotationCache()
             )

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/fieldchecks/ReflexivityFieldCheck.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/checkers/fieldchecks/ReflexivityFieldCheck.java
@@ -100,18 +100,21 @@ public class ReflexivityFieldCheck<T> implements FieldCheck<T> {
         ObjectAccessor<T> copyAccessor,
         FieldAccessor fieldAccessor
     ) {
-        Field field = fieldAccessor.getField();
-        boolean fieldIsPrimitive = fieldAccessor.fieldIsPrimitive();
-        boolean fieldIsNonNull = NonnullAnnotationVerifier.fieldIsNonnull(field, annotationCache);
-        boolean ignoreNull =
-            fieldIsNonNull ||
-            warningsToSuppress.contains(Warning.NULL_FIELDS) ||
-            nonnullFields.contains(field.getName());
-        if (fieldIsPrimitive || !ignoreNull) {
-            T left = referenceAccessor.withDefaultedField(field).get();
-            T right = copyAccessor.withDefaultedField(field).get();
-            checkReflexivityFor(left, right);
+        if (fieldAccessor.fieldIsPrimitive() && warningsToSuppress.contains(Warning.ZERO_FIELDS)) {
+            return;
         }
+
+        Field field = fieldAccessor.getField();
+        boolean nullWarningIsSuppressed = warningsToSuppress.contains(Warning.NULL_FIELDS);
+        boolean fieldIsNonNull = NonnullAnnotationVerifier.fieldIsNonnull(field, annotationCache);
+        boolean fieldIsMentionedExplicitly = nonnullFields.contains(field.getName());
+        if (nullWarningIsSuppressed || fieldIsNonNull || fieldIsMentionedExplicitly) {
+            return;
+        }
+
+        T left = referenceAccessor.withDefaultedField(field).get();
+        T right = copyAccessor.withDefaultedField(field).get();
+        checkReflexivityFor(left, right);
     }
 
     private void checkReflexivityFor(T left, T right) {

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/reflection/ClassAccessor.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/reflection/ClassAccessor.java
@@ -211,8 +211,10 @@ public class ClassAccessor<T> {
      *
      * @param enclosingType Describes the type that contains this object as a field, to determine
      *     any generic parameters it may contain.
-     * @param isSuppressed Whether reference fields are allowed to be null (a.k.a., whether
+     * @param isWarningNullSuppressed Whether reference fields must be non-null (a.k.a., whether
      *     Warnings.NULL_FIELDS is suppressed).
+     * @param isWarningZeroSuppressed Whether primitive fields must be non-0 (a.k.a., whether
+     *     Warnings.ZERO_FIELDS is suppressed).
      * @param nonnullFields Fields which are not allowed to be set to null.
      * @param annotationCache To check for any NonNull annotations.
      * @return An {@link ObjectAccessor} for an instance of T where all the fields are initialized
@@ -220,29 +222,39 @@ public class ClassAccessor<T> {
      */
     public ObjectAccessor<T> getDefaultValuesAccessor(
         TypeTag enclosingType,
-        boolean isSuppressed,
+        boolean isWarningNullSuppressed,
+        boolean isWarningZeroSuppressed,
         Set<String> nonnullFields,
         AnnotationCache annotationCache
     ) {
         Predicate<Field> canBeDefault = f ->
-            canBeDefault(f, enclosingType, isSuppressed, nonnullFields, annotationCache);
+            canBeDefault(
+                f,
+                enclosingType,
+                isWarningNullSuppressed,
+                isWarningZeroSuppressed,
+                nonnullFields,
+                annotationCache
+            );
         return buildObjectAccessor().clear(canBeDefault, prefabValues, enclosingType);
     }
 
     private boolean canBeDefault(
         Field f,
         TypeTag enclosingType,
-        boolean isSuppressed,
+        boolean isWarningNullSuppressed,
+        boolean isWarningZeroSuppressed,
         Set<String> nonnullFields,
         AnnotationCache annotationCache
     ) {
         FieldAccessor accessor = FieldAccessor.of(f);
         if (accessor.fieldIsPrimitive()) {
-            return true;
+            return !isWarningZeroSuppressed;
         }
+
         boolean isAnnotated = NonnullAnnotationVerifier.fieldIsNonnull(f, annotationCache);
         boolean isMentionedExplicitly = nonnullFields.contains(f.getName());
-        return !isSuppressed && !isAnnotated && !isMentionedExplicitly;
+        return !isWarningNullSuppressed && !isAnnotated && !isMentionedExplicitly;
     }
 
     private ObjectAccessor<T> buildObjectAccessor() {

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/reflection/RecordObjectAccessor.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/reflection/RecordObjectAccessor.java
@@ -139,7 +139,9 @@ final class RecordObjectAccessor<T> extends ObjectAccessor<T> {
     private T callRecordConstructor(List<?> params) {
         return rethrow(
             () -> constructor.newInstance(params.toArray(new Object[0])),
-            "Record: failed to invoke constructor."
+            "Record: failed to invoke constructor.\n" +
+            "  If the record does not accept 0 as a value for its fields," +
+            " consider providing valid prefab values for those fields and suppressing Warning.ZERO_FIELDS"
         );
     }
 }

--- a/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/reflection/RecordObjectAccessor.java
+++ b/equalsverifier-core/src/main/java/nl/jqno/equalsverifier/internal/reflection/RecordObjectAccessor.java
@@ -141,7 +141,7 @@ final class RecordObjectAccessor<T> extends ObjectAccessor<T> {
             () -> constructor.newInstance(params.toArray(new Object[0])),
             "Record: failed to invoke constructor.\n" +
             "  If the record does not accept 0 as a value for its fields," +
-            " consider providing valid prefab values for those fields and suppressing Warning.ZERO_FIELDS"
+            " consider providing valid prefab values for those fields and suppressing Warning.ZERO_FIELDS."
         );
     }
 }

--- a/equalsverifier-test-core/src/test/java/nl/jqno/equalsverifier/internal/checkers/FieldInspectorTest.java
+++ b/equalsverifier-test-core/src/test/java/nl/jqno/equalsverifier/internal/checkers/FieldInspectorTest.java
@@ -37,6 +37,8 @@ public class FieldInspectorTest {
         AnnotationCache annotationCache = new AnnotationCache();
 
         inspector.checkWithNull(
+            false,
+            false,
             nullFields,
             annotationCache,
             new ResetObjectForEachIterationCheck<>()

--- a/equalsverifier-test-core/src/test/java/nl/jqno/equalsverifier/internal/reflection/ClassAccessorTest.java
+++ b/equalsverifier-test-core/src/test/java/nl/jqno/equalsverifier/internal/reflection/ClassAccessorTest.java
@@ -213,6 +213,7 @@ public class ClassAccessorTest {
         ObjectAccessor<DefaultValues> objectAccessor = defaultValuesClassAccessor.getDefaultValuesAccessor(
             TypeTag.NULL,
             false,
+            false,
             new HashSet<>(),
             defaultValuesAnnotationCache
         );
@@ -228,6 +229,7 @@ public class ClassAccessorTest {
         ObjectAccessor<DefaultValues> objectAccessor = defaultValuesClassAccessor.getDefaultValuesAccessor(
             TypeTag.NULL,
             false,
+            false,
             nonnullFields,
             defaultValuesAnnotationCache
         );
@@ -241,6 +243,7 @@ public class ClassAccessorTest {
         ObjectAccessor<DefaultValues> objectAccessor = defaultValuesClassAccessor.getDefaultValuesAccessor(
             TypeTag.NULL,
             true,
+            false,
             new HashSet<>(),
             defaultValuesAnnotationCache
         );
@@ -250,11 +253,26 @@ public class ClassAccessorTest {
     }
 
     @Test
+    public void getDefaultValuesAccessor_whenZeroWarningIsSuppressed() {
+        ObjectAccessor<DefaultValues> objectAccessor = defaultValuesClassAccessor.getDefaultValuesAccessor(
+            TypeTag.NULL,
+            false,
+            true,
+            new HashSet<>(),
+            defaultValuesAnnotationCache
+        );
+        DefaultValues foo = objectAccessor.get();
+        assertNotEquals(0, foo.i);
+        // The rest is tested in getDefaultValuesObject
+    }
+
+    @Test
     public void getDefaultValuesAccessor_objectContent() {
         ClassAccessor<DefaultValues> accessor = ClassAccessor.of(DefaultValues.class, prefabValues);
         DefaultValues foo = accessor
             .getDefaultValuesAccessor(
                 TypeTag.NULL,
+                false,
                 false,
                 new HashSet<>(),
                 defaultValuesAnnotationCache

